File: app/src/main/java/com/example/androidscript/Activities/Basic/BasicBlockAdapter.java
package com.example.androidscript.Activities.Basic;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import com.example.androidscript.R;
import com.example.androidscript.UITemplate.BlockAdapter;

import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.Vector;

public class BasicBlockAdapter extends BlockAdapter<BasicViewHolder> {

    public BasicBlockAdapter(Vector<Vector<String>> content) {
        Data = content;
        this.onOrderChange = new updater() {
            @Override
            public void swap(int a, int b) {
                if (a >= 0 && b < Data.size() && a < b) {
                    Collections.swap(Data, a, b);
                    notifyDataSetChanged();
                }
            }

            @Override
            public void delete(int a) {
                if (a >= 0 && a <= Data.size() - 1) {
                    Data.remove(a);
                    notifyDataSetChanged();
                }
            }

            @Override
            public void insert() {
                notifyDataSetChanged();
            }

            @Override
            public void self(int index) {

            }
        };
    }

    @Override
    public @NotNull BasicViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view;
        switch (viewType) {
            case 0://ZeroVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_zero_var_format, parent, false);
                return new BasicViewHolder.ZeroVH(view);
            case 1://OneVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_one_var_format, parent, false);
                return new BasicViewHolder.OneVH(view);
            case 2://TwoVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_two_var_format, parent, false);
                return new BasicViewHolder.TwoVH(view);
            case 3://TwoVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_three_var_format, parent, false);
                return new BasicViewHolder.ThreeVH(view);
            case 4://FourVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_four_var_format, parent, false);
                return new BasicViewHolder.FourVH(view);
            case 5://FiveVarFormat
                view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.basic_five_var_format, parent, false);
                return new BasicViewHolder.FiveVH(view);
        }
        throw new RuntimeException("Invalid Type");
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull BasicViewHolder holder, int position) {
        holder.onBind(onOrderChange, position, this.Data);
        holder.Title.setText(Data.get(position).get(0));
        switch (Data.get(position).get(0)) {
            case "Exit"://ZeroVH
                break;
            case "Log":
                holder.Inputs[0].setHint("LogString");
                break;
            case "JumpTo":
                holder.Inputs[0].setHint("Line");
                break;
            case "Wait":
                holder.Inputs[0].setHint("ms");
                break;
            case "Call":
                holder.Inputs[0].setHint(".txt");
                break;
            case "Tag":
                holder.Inputs[0].setHint("$Var");
                break;
            case "Return":
                holder.Inputs[0].setHint("Value");
                break;
            case "ClickPic"://TwoVH
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("");
                holder.Inputs[0].setHint("Image");
                holder.Inputs[1].setHint("Ratio");
                break;
            case "Click":
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("");
                holder.Inputs[0].setHint("X");
                holder.Inputs[1].setHint("Y");
                break;
            case "CallArg":
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("");
                holder.Inputs[0].setHint(".txt");
                holder.Inputs[1].setHint("Value");
                break;
            case "IfGreater":
                holder.Title.setText("If");
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText(">");
                holder.Inputs[0].setHint("$Var");
                holder.Inputs[1].setHint("Value");
                break;
            case "IfSmaller":
                holder.Title.setText("If");
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("<");
                holder.Inputs[0].setHint("$Var");
                holder.Inputs[1].setHint("Value");
                break;
            case "Add":
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("+=");
                holder.Inputs[0].setHint("$Var");
                holder.Inputs[1].setHint("Value");
                break;
            case "Subtract":
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("-=");
                holder.Inputs[0].setHint("$Var");
                holder.Inputs[1].setHint("Value");
                break;
            case "Var":
                ((BasicViewHolder.TwoVH) holder).TitleMiddle.setText("=");
                holder.Inputs[0].setHint("$Name");
                holder.Inputs[1].setHint("Value");
                break;
            case "Check"://Three
                holder.Inputs[0].setHint("x");
                holder.Inputs[1].setHint("y");
                holder.Inputs[2].setHint("color");
                break;
            case "Swipe"://FourVH
                holder.Inputs[0].setHint("FromX");
                holder.Inputs[1].setHint("FromY");
                holder.Inputs[2].setHint("ToX");
                holder.Inputs[3].setHint("ToY");
                break;
            case "Compare"://FiveVH
                holder.Inputs[0].setHint("X1");
                holder.Inputs[1].setHint("Y1");
                holder.Inputs[2].setHint("X2");
                holder.Inputs[3].setHint("Y2");
                holder.Inputs[4].setHint("Image");
                break;
        }
    }

    @Override
    public int getItemViewType(int position) {
        switch (Data.get(position).get(0)) {
            case "Exit":
                return 0;
            case "Log":
            case "JumpTo":
            case "Wait":
            case "Call":
            case "Tag":
            case "Return":
                return 1;
            case "ClickPic":
            case "Click":
            case "CallArg":
            case "IfGreater":
            case "IfSmaller":
            case "Add":
            case "Subtract":
            case "Var":
                return 2;
            case "Check":
                return 3;
            case "Swipe":
                return 4;
            case "Compare":
                return 5;
        }
        throw new RuntimeException("Invalid " + Data.get(position).get(0));
    }

    @Override
    public int getItemCount() {
        return Data.size();
    }
}

File: app/src/main/java/com/example/androidscript/activities/basic/BasicBlockAdapter.kt
package com.example.androidscript.activities.basic

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.androidscript.R
import com.example.androidscript.activities.basic.BasicViewHolder.*
import com.example.androidscript.activities.template.BlockAdapter
import com.example.androidscript.core.Commands
import com.example.androidscript.core.Commands.getCommandLength
import java.util.*

class BasicBlockAdapter(content: Vector<Vector<String>>) : BlockAdapter<BasicViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BasicViewHolder {
        val view: View
        when (viewType) {
            0 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_zero_var_format, parent, false)
                return ZeroVH(view)
            }
            1 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_one_var_format, parent, false)
                return OneVH(view)
            }
            2 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_two_var_format, parent, false)
                return TwoVH(view)
            }
            3 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_three_var_format, parent, false)
                return ThreeVH(view)
            }
            4 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_four_var_format, parent, false)
                return FourVH(view)
            }
            5 -> {
                view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.basic_five_var_format, parent, false)
                return FiveVH(view)
            }
        }
        throw RuntimeException("Invalid view type!")
    }

    override fun onBindViewHolder(holder: BasicViewHolder, position: Int) {
        holder.onBind(onOrderChange, position, data)
        holder.title.text = data[position][0]
        when (data[position][0]) {
            Commands.EXIT -> {}
            Commands.LOG -> holder.inputs[0]!!.hint = "LogString"
            Commands.JUMP_TO -> holder.inputs[0]!!.hint = "Line"
            Commands.WAIT -> holder.inputs[0]!!.hint = "ms"
            Commands.CALL -> holder.inputs[0]!!.hint = ".txt"
            Commands.TAG -> holder.inputs[0]!!.hint = "\$Var"
            Commands.RETURN -> holder.inputs[0]!!.hint = "Value"
            Commands.CLICK_PIC -> {
                (holder as TwoVH).titleMiddle.text = ""
                holder.inputs[0]!!.hint = "Image"
                holder.inputs[1]!!.hint = "Ratio"
            }
            Commands.CLICK -> {
                (holder as TwoVH).titleMiddle.text = ""
                holder.inputs[0]!!.hint = "X"
                holder.inputs[1]!!.hint = "Y"
            }
            Commands.CALL_ARG -> {
                (holder as TwoVH).titleMiddle.text = ""
                holder.inputs[0]!!.hint = ".txt"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.IF_GREATER -> {
                holder.title.text = "If"
                (holder as TwoVH).titleMiddle.text = ">"
                holder.inputs[0]!!.hint = "\$Var"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.IF_SMALLER -> {
                holder.title.text = "If"
                (holder as TwoVH).titleMiddle.text = "<"
                holder.inputs[0]!!.hint = "\$Var"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.ADD -> {
                (holder as TwoVH).titleMiddle.text = "+="
                holder.inputs[0]!!.hint = "\$Var"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.SUBTRACT -> {
                (holder as TwoVH).titleMiddle.text = "-="
                holder.inputs[0]!!.hint = "\$Var"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.VAR -> {
                (holder as TwoVH).titleMiddle.text = "="
                holder.inputs[0]!!.hint = "\$Name"
                holder.inputs[1]!!.hint = "Value"
            }
            Commands.CHECK -> {
                holder.inputs[0]!!.hint = "x"
                holder.inputs[1]!!.hint = "y"
                holder.inputs[2]!!.hint = "color"
            }
            Commands.SWIPE -> {
                holder.inputs[0]!!.hint = "FromX"
                holder.inputs[1]!!.hint = "FromY"
                holder.inputs[2]!!.hint = "ToX"
                holder.inputs[3]!!.hint = "ToY"
            }
            Commands.COMPARE -> {
                holder.inputs[0]!!.hint = "X1"
                holder.inputs[1]!!.hint = "Y1"
                holder.inputs[2]!!.hint = "X2"
                holder.inputs[3]!!.hint = "Y2"
                holder.inputs[4]!!.hint = "Image"
            }
        }
    }

    override fun getItemViewType(position: Int) = getCommandLength(data[position][0])

    override fun getItemCount() = data.size

    init {
        data = content
        onOrderChange = object : Updater {
            override fun swap(a: Int, b: Int) {
                if (a >= 0 && b < data.size && a < b) {
                    Collections.swap(data, a, b)
                    notifyDataSetChanged()
                }
            }

            override fun delete(a: Int) {
                if (a >= 0 && a <= data.size - 1) {
                    data.removeAt(a)
                    notifyDataSetChanged()
                }
            }

            override fun insert() = notifyDataSetChanged()
            override fun self(index: Int) {}
        }
    }
}